src/clean-core/format.hh

```diff
@@ -0,0 +1,184 @@
+
+template <class T, class = std::void_t<>>
+struct has_to_string_t : std::false_type
+{
+};
+template <class T>
+struct has_to_string_t<T, std::void_t<decltype(string_view(to_string(std::declval<T>())))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_to_string = has_to_string_t<T>::value;
+
+template <class T, class = std::void_t<>>
+struct has_member_to_string_t : std::false_type
+{
+};
+template <class T>
+struct has_member_to_string_t<T, std::void_t<decltype(string_view(std::declval<T>().to_string()))>> : std::true_type
+{
+};
+template <class T>
+constexpr bool has_member_to_string = has_member_to_string_t<T>::value;
+
+struct default_formatter
+{
+    template <class T>
+    static void do_format(string_stream& ss, T const& v, string_view fmt_args)
+    {
+        if constexpr (detail::has_to_string_ss_args<T>)
+        {
+            to_string(ss, v, fmt_args);
+        }
+        else if constexpr (detail::has_to_string_args<T>)
+        {
+            if constexpr (detail::has_to_string_ss<T>)
+            {
+                if (fmt_args.empty())
+                    to_string(ss, v);
+                else
+                    ss << to_string(v, fmt_args);
+            }
+            else
+            {
+                ss << string_view(to_string(v, fmt_args));
+            }
+        }
+        else if constexpr (detail::has_to_string_ss<T>)
+        {
+            to_string(ss, v);
+        }
+        else if constexpr (detail::has_to_string<T>)
+        {
+            ss << string_view(to_string(v));
+        }
+        else if constexpr (detail::has_member_to_string<T>)
+        {
+            ss << string_view(v.to_string());
+        }
+        else
+        {
+            static_assert(cc::always_false<T>, "Type requires a to_string() method");
+        }
+    }
+};
+
+struct arg_info
+{
+    function_ptr<void(string_stream&, void const*, string_view)> do_format;
+    void const* data = nullptr;
+    string_view name;
+};
+
+template <class Formatter = default_formatter, class T>
+arg_info make_arg_info(T const& v)
+{
+    return {[](string_stream& ss, void const* data, string_view options) -> void { Formatter::do_format(ss, *static_cast<T const*>(data), options); }, &v};
+}
+
+template <class Formatter = default_formatter, class T>
+arg_info make_arg_info(format_arg<T> const& a)
+{
+    return {[](string_stream& ss, void const* data, string_view options) -> void { Formatter::do_format(ss, *static_cast<T const*>(data), options); },
+            &a.value, a.name};
+}
+
+void vformat_to(string_stream& ss, string_view fmt_str, span<arg_info> args);
+}
+
+template <class T>
+struct format_arg
+{
+    format_arg(string_view name, T const& v) : name{name}, value{v} {}
+    string_view name;
+    T const& value;
+};
+
+template <class Formatter = detail::default_formatter, class... Args>
+void format_to(string_stream& ss, string_view fmt_str, Args const&... args)
```

Comment:
The basic use-case is of course similar to `cc::format("{}", "Hello World!");`, but instead we format into a `cc::string_stream`. 
```cpp
cc::string_stream ss;
cc::format_to(ss, "{} {}!", "Hello", "World");
```
However, you are probably referring to the `Formatter` template argument :)
The reason this exists is to allow customization of formatting any type, including the builtin types.
Usually, types are formatted by their corresponding `to_string(....)` function.
However, this might not always be what you want for your use-case.
Consider a formatter that works as before for all types except for `int`, but instead of a numerical representation inserts the character sequence "int".
```cpp
struct my_formatter
{
    template <class T>
    static void do_format(string_stream& ss, T const& v, string_view fmt_args)
    {
         cc::default_formatter::do_format(ss, v, fmt_args);
    }
   static void do_format(string_stream& ss, int const& v, string_view)
    {
        ss << "int";
    }
};
```
Usage is then as follows:

```cpp
cc::string_stream ss;
cc::format_to<my_formatter>(ss, "{}", 42);
```
The result is now `"int"`.

Reply:
And if your question was about syntax for the `fmt_string`, it is essentially the same as for [fmtlib](https://fmt.dev/latest/). 

It is planned to extend the syntax slightly to enable all types (even custom ones) to use alignment options without having to re-implement the logic in their `to_string(...)`. Currently the part inside `{...}` consists of an optional id, which can be followed by `:` and some type specific options (have a look at fmtlibs [syntax page](https://fmt.dev/latest/syntax.html) for more information). The plan is to allow a second `:`, after which an additional set of alignment options can be declared.
Example: `{my_arg_name:my_type_specific_options:-<100}`.

Reply:
I mostly meant a comment in code :) But thanks!